File: web/src/api/subscriptions.ts
import { apiClient, listQuery } from './client'
import type {
  ListQuery,
  Page,
  PlanInput,
  PlanStatus,
  ServicePlan,
  Subscription,
  SubscriptionInput,
  SubscriptionStatus,
  SubscriptionUpdateInput,
} from './types'

const base = '/api/control'
const mutationHeaders = (idempotencyKey: string) => ({ 'Idempotency-Key': idempotencyKey })

export const subscriptionsApi = {
  plans: (includeArchived = false, signal?: AbortSignal) =>
    apiClient
      .request<ServicePlanWire[]>(`${base}/plans`, {
        query: { includeArchived },
        ...(signal ? { signal } : {}),
      })
      .then((items) => items.map(mapPlan)),
  publishPlan: (input: PlanInput, idempotencyKey: string, id?: string) =>
    apiClient
      .request<ServicePlanWire>(id ? `${base}/plans/${encodeURIComponent(id)}` : `${base}/plans`, {
        method: id ? 'PUT' : 'POST',
        body: input,
        headers: mutationHeaders(idempotencyKey),
      })
      .then(mapPlan),
  setPlanStatus: (id: string, status: PlanStatus, idempotencyKey: string) =>
    apiClient
      .request<ServicePlanWire>(`${base}/plans/${encodeURIComponent(id)}/status`, {
        method: 'PUT',
        body: { status },
        headers: mutationHeaders(idempotencyKey),
      })
      .then(mapPlan),
  subscriptions: (query: ListQuery, signal?: AbortSignal) =>
    apiClient
      .request<Page<SubscriptionWire>>(`${base}/subscriptions`, {
        query: listQuery(query),
        ...(signal ? { signal } : {}),
      })
      .then((page) => ({ ...page, items: page.items.map(mapSubscription) })),
  createSubscription: (input: SubscriptionInput, idempotencyKey: string) =>
    apiClient
      .request<SubscriptionWire>(`${base}/subscriptions`, {
        method: 'POST',
        body: input,
        headers: mutationHeaders(idempotencyKey),
      })
      .then(mapSubscription),
  updateSubscription: (id: string, input: SubscriptionUpdateInput, idempotencyKey: string) =>
    apiClient
      .request<SubscriptionWire>(`${base}/subscriptions/${encodeURIComponent(id)}`, {
        method: 'PUT',
        body: input,
        headers: mutationHeaders(idempotencyKey),
      })
      .then(mapSubscription),
  setSubscriptionStatus: (
    id: string,
    status: SubscriptionStatus,
    expectedUpdatedAt: string,
    idempotencyKey: string,
  ) =>
    apiClient
      .request<SubscriptionWire>(`${base}/subscriptions/${encodeURIComponent(id)}/status`, {
        method: 'PUT',
        body: { status, expectedUpdatedAt },
        headers: mutationHeaders(idempotencyKey),
      })
      .then(mapSubscription),
}

interface PlanRouteWire {
  model_id: string
  model_name?: string
  resource_pool_id: string
  resource_pool_name?: string
  resource_pool_slug?: string
  provider_name?: string
}

interface PlanVersionWire {
  id: string
  version: number
  token_quota: number
  validity_days: number
  concurrency_limit: number
  rpm_limit?: number
  tpm_limit?: number
  routes: PlanRouteWire[]
  created_at: string
}

interface ServicePlanWire {
  id: string
  slug: string
  name: string
  description: string
  kind: ServicePlan['kind']
  status: ServicePlan['status']
  current_version?: PlanVersionWire
  active_subscription_count: number
  created_at: string
  updated_at: string
}

interface SubscriptionWire {
  id: string
  user_id: string
  member_email: string
  member_name: string
  service_plan_id: string
  service_plan_version_id: string
  service_plan_name: string
  plan_kind: Subscription['planKind']
  plan_version: number
  status: Subscription['status']
  granted_tokens: number
  balance_tokens: number
  starts_at: string
  expires_at: string
  notes: string
  concurrency_limit: number
  rpm_limit?: number
  tpm_limit?: number
  routes: PlanRouteWire[]
  suspended_at?: string
  canceled_at?: string
  created_at: string
  updated_at: string
}

function mapPlan(plan: ServicePlanWire): ServicePlan {
  return {
    id: plan.id,
    slug: plan.slug,
    name: plan.name,
    description: plan.description,
    kind: plan.kind,
    status: plan.status,
    ...(plan.current_version
      ? {
          currentVersion: {
            id: plan.current_version.id,
            version: plan.current_version.version,
            tokenQuota: plan.current_version.token_quota,
            validityDays: plan.current_version.validity_days,
            concurrencyLimit: plan.current_version.concurrency_limit,
            ...(plan.current_version.rpm_limit !== undefined
              ? { rpmLimit: plan.current_version.rpm_limit }
              : {}),
            ...(plan.current_version.tpm_limit !== undefined
              ? { tpmLimit: plan.current_version.tpm_limit }
              : {}),
            routes: plan.current_version.routes.map((route) => ({
              modelId: route.model_id,
              modelName: route.model_name ?? route.model_id,
              resourcePoolId: route.resource_pool_id,
              resourcePoolName: route.resource_pool_name ?? route.resource_pool_id,
              resourcePoolSlug: route.resource_pool_slug ?? '',
              providerName: route.provider_name ?? '',
            })),
            createdAt: plan.current_version.created_at,
          },
        }
      : {}),
    activeSubscriptionCount: plan.active_subscription_count,
    createdAt: plan.created_at,
    updatedAt: plan.updated_at,
  }
}

function mapSubscription(value: SubscriptionWire): Subscription {
  return {
    id: value.id,
    userId: value.user_id,
    memberEmail: value.member_email,
    memberName: value.member_name,
    servicePlanId: value.service_plan_id,
    servicePlanVersionId: value.service_plan_version_id,
    servicePlanName: value.service_plan_name,
    planKind: value.plan_kind,
    planVersion: value.plan_version,
    status: value.status,
    grantedTokens: value.granted_tokens,
    balanceTokens: value.balance_tokens,
    startsAt: value.starts_at,
    expiresAt: value.expires_at,
    notes: value.notes,
    concurrencyLimit: value.concurrency_limit,
    ...(value.rpm_limit !== undefined ? { rpmLimit: value.rpm_limit } : {}),
    ...(value.tpm_limit !== undefined ? { tpmLimit: value.tpm_limit } : {}),
    routes: value.routes.map((route) => ({
      modelId: route.model_id,
      modelName: route.model_name ?? route.model_id,
      resourcePoolId: route.resource_pool_id,
      resourcePoolName: route.resource_pool_name ?? route.resource_pool_id,
      resourcePoolSlug: route.resource_pool_slug ?? '',
      providerName: route.provider_name ?? '',
    })),
    ...(value.suspended_at ? { suspendedAt: value.suspended_at } : {}),
    ...(value.canceled_at ? { canceledAt: value.canceled_at } : {}),
    createdAt: value.created_at,
    updatedAt: value.updated_at,
  }
}

File: internal/controlapi/subscriptions.go
package controlapi

import (
	"net/http"
	"strconv"
	"time"

	"github.com/go-chi/chi/v5"
	"github.com/google/uuid"
	"github.com/luckymaomi/llmgateway/internal/httpserver"
	"github.com/luckymaomi/llmgateway/internal/subscription"
)

type planInput struct {
	Name             string                `json:"name"`
	Description      string                `json:"description"`
	Kind             subscription.PlanKind `json:"kind"`
	TokenQuota       int64                 `json:"tokenQuota"`
	ValidityDays     int32                 `json:"validityDays"`
	ConcurrencyLimit int32                 `json:"concurrencyLimit"`
	RPMLimit         *int32                `json:"rpmLimit"`
	TPMLimit         *int64                `json:"tpmLimit"`
	Routes           []struct {
		ModelID        uuid.UUID `json:"modelId"`
		ResourcePoolID uuid.UUID `json:"resourcePoolId"`
	} `json:"routes"`
}

func (a *API) listPlans(w http.ResponseWriter, r *http.Request) {
	includeArchived, _ := strconv.ParseBool(r.URL.Query().Get("includeArchived"))
	items, err := a.subscriptions.ListPlans(r.Context(), principalFromContext(r.Context()), includeArchived)
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	writeData(w, http.StatusOK, items)
}

func (a *API) publishPlan(w http.ResponseWriter, r *http.Request) {
	var input planInput
	if err := decodeJSON(w, r, &input); err != nil {
		writeDecodeError(w, r, err)
		return
	}
	planID := uuid.Nil
	if value := chi.URLParam(r, "planID"); value != "" {
		parsed, err := uuid.Parse(value)
		if err != nil {
			writeDecodeError(w, r, err)
			return
		}
		planID = parsed
	}
	routes := make([]subscription.PlanRoute, 0, len(input.Routes))
	for _, route := range input.Routes {
		routes = append(routes, subscription.PlanRoute{ModelID: route.ModelID, ResourcePoolID: route.ResourcePoolID})
	}
	mutation, ok := subscriptionMutationRequest(w, r)
	if !ok {
		return
	}
	item, err := a.subscriptions.PublishPlan(r.Context(), principalFromContext(r.Context()), subscription.PlanDraft{
		ID: planID, Name: input.Name, Description: input.Description, Kind: input.Kind,
		TokenQuota: input.TokenQuota, ValidityDays: input.ValidityDays, ConcurrencyLimit: input.ConcurrencyLimit,
		RPMLimit: input.RPMLimit, TPMLimit: input.TPMLimit, Routes: routes,
	}, mutation)
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	status := http.StatusCreated
	if planID != uuid.Nil {
		status = http.StatusOK
	}
	writeData(w, status, item)
}

func (a *API) setPlanStatus(w http.ResponseWriter, r *http.Request) {
	id, ok := pathUUID(w, r, "planID")
	if !ok {
		return
	}
	var input struct {
		Status subscription.PlanStatus `json:"status"`
	}
	if err := decodeJSON(w, r, &input); err != nil {
		writeDecodeError(w, r, err)
		return
	}
	mutation, ok := subscriptionMutationRequest(w, r)
	if !ok {
		return
	}
	item, err := a.subscriptions.SetPlanStatus(r.Context(), principalFromContext(r.Context()), id, input.Status, mutation)
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	writeData(w, http.StatusOK, item)
}

func (a *API) listSubscriptions(w http.ResponseWriter, r *http.Request) {
	query := parseListQuery(r)
	var userID *uuid.UUID
	if query.UserID != "" {
		value, err := uuid.Parse(query.UserID)
		if err != nil {
			writeDecodeError(w, r, err)
			return
		}
		userID = &value
	}
	page, err := a.subscriptions.ListSubscriptions(r.Context(), principalFromContext(r.Context()), subscription.Query{UserID: userID, Search: query.Search, Status: query.Status, Offset: query.offset(), Size: int32(query.PageSize)})
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	writeData(w, http.StatusOK, pageView[subscription.Subscription]{Items: page.Items, Total: page.Total, Page: query.Page, PageSize: query.PageSize})
}

func (a *API) createSubscription(w http.ResponseWriter, r *http.Request) {
	var input struct {
		UserID        uuid.UUID `json:"userId"`
		ServicePlanID uuid.UUID `json:"servicePlanId"`
		GrantedTokens int64     `json:"grantedTokens"`
		StartsAt      time.Time `json:"startsAt"`
		ExpiresAt     time.Time `json:"expiresAt"`
		Notes         string    `json:"notes"`
	}
	if err := decodeJSON(w, r, &input); err != nil {
		writeDecodeError(w, r, err)
		return
	}
	mutation, ok := subscriptionMutationRequest(w, r)
	if !ok {
		return
	}
	item, err := a.subscriptions.CreateSubscription(r.Context(), principalFromContext(r.Context()), subscription.NewSubscription{UserID: input.UserID, ServicePlanID: input.ServicePlanID, GrantedTokens: input.GrantedTokens, StartsAt: input.StartsAt, ExpiresAt: input.ExpiresAt, Notes: input.Notes}, mutation)
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	writeData(w, http.StatusCreated, item)
}

func (a *API) updateSubscription(w http.ResponseWriter, r *http.Request) {
	id, ok := pathUUID(w, r, "subscriptionID")
	if !ok {
		return
	}
	var input struct {
		GrantedTokens     int64     `json:"grantedTokens"`
		StartsAt          time.Time `json:"startsAt"`
		ExpiresAt         time.Time `json:"expiresAt"`
		Notes             string    `json:"notes"`
		ExpectedUpdatedAt time.Time `json:"expectedUpdatedAt"`
	}
	if err := decodeJSON(w, r, &input); err != nil {
		writeDecodeError(w, r, err)
		return
	}
	mutation, ok := subscriptionMutationRequest(w, r)
	if !ok {
		return
	}
	item, err := a.subscriptions.UpdateSubscription(r.Context(), principalFromContext(r.Context()), subscription.SubscriptionChange{ID: id, GrantedTokens: input.GrantedTokens, StartsAt: input.StartsAt, ExpiresAt: input.ExpiresAt, Notes: input.Notes, ExpectedUpdatedAt: input.ExpectedUpdatedAt}, mutation)
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	writeData(w, http.StatusOK, item)
}

func (a *API) setSubscriptionStatus(w http.ResponseWriter, r *http.Request) {
	id, ok := pathUUID(w, r, "subscriptionID")
	if !ok {
		return
	}
	var input struct {
		Status            subscription.SubscriptionStatus `json:"status"`
		ExpectedUpdatedAt time.Time                       `json:"expectedUpdatedAt"`
	}
	if err := decodeJSON(w, r, &input); err != nil {
		writeDecodeError(w, r, err)
		return
	}
	mutation, ok := subscriptionMutationRequest(w, r)
	if !ok {
		return
	}
	item, err := a.subscriptions.SetSubscriptionStatus(r.Context(), principalFromContext(r.Context()), id, input.Status, input.ExpectedUpdatedAt, mutation)
	if err != nil {
		a.writeSubscriptionError(w, r, err)
		return
	}
	writeData(w, http.StatusOK, item)
}

func subscriptionMutationRequest(w http.ResponseWriter, r *http.Request) (subscription.MutationRequest, bool) {
	idempotencyKey, err := uuid.Parse(r.Header.Get("Idempotency-Key"))
	if err != nil || idempotencyKey == uuid.Nil {
		writeProblem(w, r, problem{Status: http.StatusBadRequest, Code: "invalid_idempotency_key", Message: "Idempotency-Key must be a UUID.", Stage: "subscription"})
		return subscription.MutationRequest{}, false
	}
	return subscription.MutationRequest{IdempotencyKey: idempotencyKey, RequestID: httpserver.RequestIDFromContext(r.Context())}, true
}
